platform/osgi/platform-osgi-configadmin/src/main/java/org/apache/felix/cm/impl/ConfigurationManager.java

```diff
@@ -230,11 +240,14 @@ public void start(BundleContext bundleContext) {
 
     // set up the location (might throw IllegalArgumentException)
     try {
-      // the only major change codice made to this class is wrapping cache persistence with
-      // encryption
-      PersistenceManager fpm =
-          new EncryptingPersistenceManager(
-              new FilePersistenceManager(bundleContext, bundleContext.getProperty(CM_CONFIG_DIR)));
+      // (CODICE) the only major change made to this class is wrapping cache persistence with
```

Comment:
❓ Wouldn't it have been easier to keep the lock inside the DelegatingPersistenceManager and acquire the readlock in its delete() and store() method only? That would have concentrated everything in one single place.

Reply:
❗️  Doing this would make us vulnerable to deadlock. It's unlikely but it **could** happen, as follows:
1. Some configuration gets submitted for persistence, which will be referred to as the **config thread**. 
2. The **config thread** acquires a _write_ lock on the separate reentrant lock in _Felix's_ `CachingPersistentManagerProxy` class, in the `store()` method. 
3. Before the **config thread** can drill down the `PersistenceManager` layers to acquire a read lock on _our_ `configDeadZoneLock` (the stamped lock), our service appears and acquires a write lock on _our_ `configDeadZoneLock` (see `ConfigInterceptingCustomizer` in the `DelegatingPersistenceManager`). This will be called the **tracker thread** since the `ServiceTracker` supports this kind of callback interception. 
4. The **tracker thread** calls the plugin's `initialize()` method which turns around and queries config admin using `listConfigurations()`. This will trigger queries on all the `CachingPersistentManagerProxy`s, at least one of which won't be queryable until **config thread** releases the write lock. 
4. Because **config thread** is waiting on **tracker thread** and vice versa, you get deadlock.

Reply:
ok. This is what I suggested that the call to initialize() should pass the config objects to initialize to avoid that callback into cfg admin. That way, you are in full control in one place of the locks instead of having them dealt with in multiple places which increases complexity.